cmake/SYCL.cmake

```
@@ -0,0 +1,71 @@
+#Macro to find a sycl implimentation
+#Called without arguments will try to find a compiler appropriate sycl
+#With one argument finds that specific implimentation
+#With 2 arguments is as above but second argument is treated like version
+#Not happy with this yet - 
+# - should be able to accept lists of implementaions and versions
+# - and target architectures CPU,GPU etc
+macro(find_sycl)
+    set(SYCL_FOUND FALSE )
+    if (${ARGC} EQUAL 0)
+        #If intel compiler try to find the intel 
+        if (CMAKE_CXX_COMPILER_ID MATCHES "Intel")
+            #TODO IntelDPCPP is deprecated remove
+            list(APPEND candidates "IntelSYCL" "IntelDPCPP")
+            foreach(candidate IN LISTS candidates)
+                find_package(${candidate} QUIET)
+                if(${candidate}_FOUND)
+                    set(SYCL_FOUND TRUE )
+                    set(SYCL_IMPLEMENTATION ${candidate} )
+                    break()
+                endif()
+            endforeach()
+        else()
+            #not intel try one of these
+            list(APPEND candidates "AdaptiveCPP" "hipSYCL")
+            list(APPEND versions   ""            "0.9.4")
+            foreach(candidate version IN ZIP_LISTS candidates versions)
+                find_package(${candidate} ${version} QUIET)
+                if(${candidate}_FOUND)
+                    set(SYCL_FOUND TRUE )
+                    set(SYCL_IMPLEMENTATION ${candidate} )
+                    break()
+                endif()
+            endforeach()
+        endif()
+    elseif (${ARGC} EQUAL 1)
+        find_package(${ARGV0} REQUIRED)
+        if(${ARGV0}_FOUND)
+            set(SYCL_FOUND TRUE )
+            set(SYCL_IMPLEMENTATION ${ARGV0} )
+        endif()
+    elseif (${ARGC} EQUAL 2)
+        find_package(${ARGV0} ${ARGV1} REQUIRED)
+        if(${ARGV0}_FOUND)
+            set(SYCL_FOUND TRUE )
+            set(SYCL_IMPLEMENTATION ${ARGV0} )
+        endif()
+    endif()
+    if (NOT COMMAND add_sycl_to_target) 
+        function(add_sycl_to_target)
+        endfunction()
```

```suggestion
        function(add_sycl_to_target)
            cmake_parse_arguments(
                PARSE_ARGV 0
                ARGS ""
                "TARGET"
                "SOURCES"
            )
            set_property(SOURCE ${ARGS_SOURCES} APPEND PROPERTY COMPILE_OPTIONS "-fsycl")
            target_link_options(${ARGS_TARGET} PRIVATE "-fsycl")
        endfunction()
```

Might be missing something but shouldn't this function act like the hipsycl version? Possibly more flags necessary than suggested.

---

IntelDPCPP doesn't need/define this function - having an empty version is just a hack to prevent some if/else clauses in the rest of the CMake files.

---

Most SYCL implementations define this function themselves. The only one that doesn't to my my knowledge is Intel and in that case the compiler doesn't need any additional flags to turn on SYCL.